### PR TITLE
First draft of The Turing Completeness Problem

### PR DESCRIPTION
**Pull request by dfellis**

### CODE REVIEW

File: src/the-turing-completeness-problem.md [hunk 1]
```diff
@@ -0,0 +1,134 @@
+# The Turing-Completeness Problem
+
+**22 October 2020 | Luis F. De Pombo, David Ellis**
+
+The problem of Turing-Completeness is not a problem with the work of Alan Turing or Alonzo Church[1]. It is a problem of an implication of their work and our industry's lack of recognition of this problem, and it has been exacerbated by the actual computing mechanisms moving further and further away from the simple "single tape" Turing Machine model of computation that our languages are founded upon.
```
Comment (depombo) [inline src/the-turing-completeness-problem.md:5]: This feels roundabout and confusing because it doesn't tell me what is the actual problem with Turing Completeness. Maybe something more straight forward that jumps straight into how it's not possible to know if the code will halt and as a result execution is unpredictable.

File: src/the-turing-completeness-problem.md [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+The following snippet of C code succinctly contains the problems of Turing completeness complicating modern development:
+
+```c
+while (node) {
+  doSomethingWith(node);
+  node = node->next;
+}
+```
+
+1. *Automatic Parallelization* - It's not possible to automatically parallelize this code. The number of nodes is not clear to the compiler for it to potentially bunch those calls across threads, and it's also unclear if the function call mutates the node and adds/removes nodes to process.
```
Comment (depombo) [inline src/the-turing-completeness-problem.md:16]: Feels to me like 1 is a consequence of 2 so their order should be inverted?

File: src/the-turing-completeness-problem.md [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+2. *Halting Problem* - It's not possible to know if the code will halt, even ignoring the idea that `doSomethingWith` may always add a new `node` to the one it is processing, there's nothing stopping a node from pointing at itself and never halting that way.
+
+Automatic parallelization is only a problem because we want/need to parallelize our code. Until very recently this was a niche need as computing power had been increasing exponentially for decades and has only flattened in the past decade[2]. If this was still the case it is likely that it would have remained a niche problem that only experts need to worry about, but with multicore now the norm and unlikely to go away[3], it has become a problem for the average developer, as well.
```
Comment (depombo) [inline src/the-turing-completeness-problem.md:19]: 2nd sentence would read better if it is expanded and also mentioned Moore's Law directly

Comment (depombo) [reply src/the-turing-completeness-problem.md:19]: what would [3] quote?

File: src/the-turing-completeness-problem.md [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+The issue with the Halting Problem itself is much more contentious. Turing demonstrated that arbitrary computability requires that the possibility of an algorithm to never halt be there, and so it has been tolerated as an unfortunate reality that must simply be dealt with. However, we posit that no *practially useful* computation will take forever because its result cannot affect our finite reality, so programmers have already been finding an imperfect subset of Turing completeness for their own problem domains to be able to actually produce valuable calculations for their clients.
```
Comment (depombo) [inline src/the-turing-completeness-problem.md:21]: business use cases instead of finite reality? also clients instead of users?

File: src/the-turing-completeness-problem.md [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+In a single-threaded world, this problem has been mostly a small annoyance -- find the section of code that the processor is executing over and over again, study how it entered this state, and figure out how to prevent that state from occurring again -- or abort on detection of said state. But in a multi-threaded, multi-machined cluster computing world, even detecting an infinite loop that spans multiple machines is a challenge, let alone gathering the data you need to debug the cause of this invalid state, further exacerbated by the non-determinism inherent to multiple interacting Turing machines not executing in lockstep with each other making reproduction of the problem even more difficult.
```
Comment (depombo) [inline src/the-turing-completeness-problem.md:23]: This feels like a stretch in that it implies that infinite loops running on a distributed environment are common and it is not such a pervasive problem in reality. Maybe this should talk about the implications of debugging code that does not have have predictable execution and how the surface area for developer mistakes is a lot larger

File: src/the-turing-completeness-problem.md [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+Solving the Halting Problem, or rather side-stepping it by defining a barely-Turing-incomplete subset, also makes the automatic parallelization story possible, because now the compiler can work with code that it can determine predictable properties of -- where and how it could parallelize the work and by having the ability to model its finite execution time also whether or not the costs of parallelization are less than the gains in doing so under various conditions.
+
+This is the thesis of Alan, that with minor constraints on the language that we believe the vast majority of developers will not notice, or at least won't mind, we can be sure that the code you write always halts, we can make models for runtime estimates at compile time that only require the size and "shape" of the data to get an answer and can therefore determine not only how to parallelize your code, but whether or not it makes sense given the data being operated on. We have not yet accomplished all of these claims, but this is now only a matter of time and effort. :)
```
Comment (depombo) [inline src/the-turing-completeness-problem.md:27]: this is great

File: src/the-turing-completeness-problem.md [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+The root of most of the problems with the example C code is involving ambiguity in the data structure and the behavior of the user-defined function called. The other problem is that the while loop is an unbounded looping construct that cannot be reasoned about except "the body of the while loop is executed zero or more times." In Alan, *arbitrary* looping and recursion are disallowed, and all data is built on top of data types with knowable constraints, no arbitrary pointer logic allowed.
```
Comment (depombo) [inline src/the-turing-completeness-problem.md:29]: I know what you mean by "no arbitrary pointer logic allowed" in this context, but it could be misconstrued or taking out of context

File: src/the-turing-completeness-problem.md [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+fn example(file1: string, file2: string): Array<Array<int64>> {
+  const matrix1 = loadFile(file1).parseMatrix()
+  const matrix2 = loadFile(file2).parseMatrix()
+  return matrix1 * matrix2
+}
+```
+
+Not only is the matrix multiplication a trivially parallelizable operation that likely will be parallelized for any large-ish matrix, but reading the files from disk and then parsing their contents may be done in parallel before the matrix multiplication because the performance benefit is there to do so.
+
+To accomplish this, we need something different from a classic Turing tape to compile to. We call this the Alan Graphcode: a bytecode format with graph annotations. A dependency graph of operations to perform, with some of these also having a subgraph of operations to perform, provides us with the information we need to know which operations are safe to execute in parallel, coupled with (implied) operation execution time estimates to determine if it makes sense to do so.
```
Comment (depombo) [inline src/the-turing-completeness-problem.md:51]: that is guaranteed to have no cycles?

File: src/the-turing-completeness-problem.md [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+This seems incompatible with any sort of looping or branching behavior at first glance, though, since a dependency graph of operations that must all be executed eventually appears to preclude the behavior of JUMP operations like the classic JNZ (Jump if Not Zero). If there is no singular "forward" or "backwards" in a tree, how can these things work at all? This is *not* what Alan restricts users from writing -- `if` statements exist in Alan and even limited versions of classic looping constructs are in a standard library[4]. In fact we will quickly demonstrate that this representation can be made Turing Complete and is therefore a lossless representation that gains information to do parallelization.
```
Comment (depombo) [inline src/the-turing-completeness-problem.md:53]: This paragraph needs a subsection and/or an introduction for how the rest of the post proceeds to explain how controlled iteration and recursion are allowed via `@std/seq` before drawing the parallel to JNZ and so on. Also there should be a url with more information about JNZ for people that are not familiar with it / assembly.

File: src/the-turing-completeness-problem.md (hunk 1 above)
Comment (dfellis) [reply src/the-turing-completeness-problem.md:5]: Yeah, I probably want to completely redo this paragraph, but I did want to make it clear that I'm not critiquing their work.

File: src/the-turing-completeness-problem.md (hunk 2 above)
Comment (dfellis) [reply src/the-turing-completeness-problem.md:16]: So I did it in this order because people see (1) as actually something worthwhile and (2) as just an academic problem that doesn't matter in reality, and then the explanation makes it clearer how (1) is because of (2). Being in reverse order from our perspective was on purpose since I think that's the ordering people will be approaching this article.

File: src/the-turing-completeness-problem.md (hunk 3 above)
Comment (dfellis) [reply src/the-turing-completeness-problem.md:19]: So `[2]` was just going to be the single-core performance of processors flatlining. `[3]` was going to be a link to the fundamental physics involved due to the natural capacitance of Silicon and Silicon Dioxide putting an upper limit on frequency with respect to chip size, period.

File: src/the-turing-completeness-problem.md (hunk 4 above)
Comment (dfellis) [reply src/the-turing-completeness-problem.md:21]: No flair for the dramatic? ;) I can make that change, but I liked hammering it in that we're mortal and for our work to be useful it has to fit within our own timescales.

Switching clients to users is fine, but I went with clients since that could also include people who are writing code for business entities and not individual people. But that's a minor quibble and I admit that "clients" sounds more stilted.

File: src/the-turing-completeness-problem.md (hunk 5 above)
Comment (dfellis) [reply src/the-turing-completeness-problem.md:23]: I don't think it's implying that it's any more common than infinite loops on an individual machine? Just because engineers have learned patterns of code to avoid because they're more likely to fail doesn't make the problem less of a reality, and it *is* significantly more difficult to deal with in a distributed case than a single-threaded case.

File: src/the-turing-completeness-problem.md (hunk 7 above)
Comment (dfellis) [reply src/the-turing-completeness-problem.md:29]: What do you mean?

File: src/the-turing-completeness-problem.md (hunk 8 above)
Comment (dfellis) [reply src/the-turing-completeness-problem.md:51]: Sure.

File: src/the-turing-completeness-problem.md (hunk 9 above)
Comment (dfellis) [reply src/the-turing-completeness-problem.md:53]: I agree on the missing citation. I don't know how to address your first point, though. How would you introduce that complexity (with the resource consumption model of the `seq` counter) *before* introducing the simpler form?

File: src/the-turing-completeness-problem.md (hunk 4 above)
Comment (depombo) [reply src/the-turing-completeness-problem.md:21]: I actually like the more metaphysical phrasing, but let's expand on it more? It feels like it is coming a bit from the blue right now

File: src/the-turing-completeness-problem.md (hunk 5 above)
Comment (depombo) [reply src/the-turing-completeness-problem.md:23]: I know that is not the case, but that is how it read to me. I do think there is a lot more value in talking about the implications of unpredictable execution and how the problem space for potential bugs explodes because of it.

File: src/the-turing-completeness-problem.md (hunk 7 above)
Comment (depombo) [reply src/the-turing-completeness-problem.md:29]: hmm maybe I misunderstood. I thought by pointer _logic_ you meant goto statements?

File: src/the-turing-completeness-problem.md (hunk 9 above)
Comment (depombo) [reply src/the-turing-completeness-problem.md:53]: We don't necessarily have to dive into `@std/seq` though I think it might be better. I just think prefacing how we do controlled recursion seems like it is necessary context before diving into JNZ

File: src/the-turing-completeness-problem.md (hunk 7 above)
Comment (dfellis) [reply src/the-turing-completeness-problem.md:29]: Nope, I mean data pointers here. You can't define data structures in Alan with arbitrary pointers.

File: src/the-turing-completeness-problem.md [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+This seems incompatible with any sort of looping or branching behavior at first glance, though, since a dependency graph of operations that must all be executed eventually appears to preclude the behavior of JUMP operations like the classic [JNZ (Jump if Not Zero)](https://www.aldeid.com/wiki/X86-assembly/Instructions/jnz). If there is no singular "forward" or "backwards" in a tree, how can these things work at all? This is *not* what Alan restricts users from writing -- `if` statements exist in Alan and even [guaranteed halting versions of classic looping constructs are in a standard library](https://docs.alan-lang.org/std_seq.html) that provide a predictable "worst case" execution time based on a maximum iteration count allowed. We will demonstrate that this graph-based representation can be made Turing Complete and is therefore a lossless representation that gains information to make automatic parallelization possible. Our further constraints placed on it are to make the question of when such parallelization should be done answerable.
```
Comment (depombo) [inline src/the-turing-completeness-problem.md:53]: the "We will demonstrate" makes me think that we are about to show a mathematical proof so maybe reword that?

File: src/the-turing-completeness-problem.md [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+  var d = op5(a);
+} while (d);
+op6(a);
+```
+
+The equivalent in a graph representation would be:
+
+![Graph Example with do-while loop](./graph-loop.png)
+
+Here a DOWHILE opcode runs the loop sub-graph over and over until the expected condition is met. So with these two special opcodes COND and DOWHILE we can reproduce the behavior of JNZ's forwards and backwards jumping and we can enter infinite loops with DOWHILE, meaning we should be Turing Complete. Representing something like [Duff's Device](https://en.wikipedia.org/wiki/Duff%27s_device) would be very difficult but not impossible here -- it just wouldn't make much sense because while the performance could be the same, the executable in this representation would be significantly larger, undoing it's main value.
```
Comment (depombo) [inline src/the-turing-completeness-problem.md:97]: not sure what the significance of this paragraph is wrt to the whole or why we are diving so deep into DOWHILE

File: src/the-turing-completeness-problem.md [hunk 12]
```diff
@@ -0,0 +1,134 @@
+# The Turing-Completeness Problem
+
+**22 October 2020 | Luis F. De Pombo, David Ellis**
+
+The work of Alan Turing and Alonzo Church on [computability](https://en.wikipedia.org/wiki/Computability) demonstrated that some things can never be computed, and some cannot be determined ahead of time whether or not they can be computed; the [Halting Problem](https://en.wikipedia.org/wiki/Halting_problem). This problem is not an issue with what they have shown, but how the software development process is affected by it and our industry's lack of recognition of the problem their results have put onto software developers. Further, it has been exacerbated by the actual computing mechanisms moving away from the simple "single tape" Turing Machine model of computation that most programming languages are founded upon.
```
[the remark this answers is not in the document]
Comment (dfellis) [reply src/the-turing-completeness-problem.md:5]: I even meant basic multithreading and NUMA architectures. :) Should I mention all three?

File: src/the-turing-completeness-problem.md [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+2. *Halting Problem* - It's not possible to know if the code will halt, even ignoring the idea that `doSomethingWith` may always add a new `node` to the one it is processing, there's nothing stopping a node from pointing at itself and never halting that way.
+
+Automatic parallelization is only a problem because we want/need to parallelize our code. Until very recently this was a niche need as [computing power had been increasing exponentially for decades and has only flattened in the past decade](https://www.karlrupp.net/2015/06/40-years-of-microprocessor-trend-data/). If this was still the case it is likely that it would have remained a niche problem that only experts need to worry about, but with multicore now the norm and unlikely to go away as [signal propagation itself becomes the limiting factor](http://www.fisgeo.unipg.it/luca.gammaitoni/fisinfo/documenti-fisici/physical-limits-silicon.pdf), it has become a problem for the average developer, as well.
```
[the remark this answers is not in the document]
Comment (dfellis) [reply src/the-turing-completeness-problem.md:19]: :+1:

File: src/the-turing-completeness-problem.md [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+
+The issue with the Halting Problem itself is much more contentious. Turing demonstrated that arbitrary computability requires that compute that never halts, and that you can't predict that ahead of time, and so it has been tolerated as an unfortunate reality that must simply be dealt with. However, we posit that no *practially useful* computation will take forever because its result cannot affect the finite reality of our lives, so programmers have already been finding an imperfect subset of Turing completeness for their own problem domains in an ad-hoc intuitive sense to be able to actually produce valuable calculations for their users.
```
[the remark this answers is not in the document]
Comment (dfellis) [reply src/the-turing-completeness-problem.md:21]: That's a lot better :)

File: src/the-turing-completeness-problem.md [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+
+Solving the Halting Problem, or rather side-stepping it by defining a barely-Turing-incomplete subset, also makes the automatic parallelization story possible, because now the compiler can work with code that it can determine predictable properties of -- where and how it could parallelize the work and by having the ability to model its finite execution time also whether or not the costs of parallelization are less than the gains in doing so under various conditions.
```
[the remark this answers is not in the document]
Comment (dfellis) [reply src/the-turing-completeness-problem.md:25]: Will need to restructure the beginning, too. The use of an em-dash won't work with it split.

File: src/the-turing-completeness-problem.md [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+
+This is the thesis of Alan, that with minor constraints on the language that we believe the vast majority of developers will not notice, or at least won't mind, we can be sure that the code you write always halts, we can make models for runtime estimates at compile time that only require the size and "shape" of the data to get an answer and can therefore determine not only how to parallelize your code, but whether or not it makes sense given the data being operated on. We have not yet accomplished all of these claims, but this is now only a matter of time and effort. :)
```
[the remark this answers is not in the document]
Comment (dfellis) [reply src/the-turing-completeness-problem.md:27]: I'll need to think about this.

File: src/the-turing-completeness-problem.md (hunk 10 above)
Comment (dfellis) [reply src/the-turing-completeness-problem.md:53]: But we are demonstrating that we can replicate the behavior of jump instructions with instructions that execute sub-graphs?

File: src/the-turing-completeness-problem.md (hunk 11 above)
Comment (dfellis) [reply src/the-turing-completeness-problem.md:97]: So, this explaining how the graph-based model can be just as Turing complete as the traditional model, and heads off a potential objection with respect to Duff's Device, by pointing out that we can replicate the behavior but not the compactness of the representation of that behavior.

I would be fine with dropping the piece on Duff's Device, but I expect it to be an objection on any forum we post this to, so that's why I put it in.

File: src/the-turing-completeness-problem.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (depombo) [reply src/the-turing-completeness-problem.md:5]: yes! it is not obvious

File: src/the-turing-completeness-problem.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (depombo) [reply src/the-turing-completeness-problem.md:25]: this seems like a run-on even with the em-dash? I do like the contents of the sentence. I will leave this up to you since you are a native english speaker and better writer overall so feel free to disregard

File: src/the-turing-completeness-problem.md (hunk 10 above)
Comment (depombo) [reply src/the-turing-completeness-problem.md:53]: true, it just doesn't seem extensive. I think it's fine. more of a nit come think of it and we can leave as-is

File: src/the-turing-completeness-problem.md (hunk 11 above)
Comment (depombo) [reply src/the-turing-completeness-problem.md:97]: > So, this explaining how the graph-based model can be just as Turing complete as the traditional model
I would love to see something as clear as this in the beginning of the section

Comment (dfellis) [reply src/the-turing-completeness-problem.md:97]: > We will demonstrate that this graph-based representation can be made Turing Complete and is therefore a lossless representation that gains information to make automatic parallelization possible.

This wasn't clear?

Comment (depombo) [reply src/the-turing-completeness-problem.md:97]: The last one you quoted has less of a "punch" or it's harder to parse just because it tries to bring two really complex and important points home in the same sentence.

File: src/the-turing-completeness-problem.md [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+fn example(file1: string, file2: string): Array<Array<int64>> {
+  const matrix1 = loadFile(file1).parseMatrix()
+  const matrix2 = loadFile(file2).parseMatrix()
+  return matrix1 * matrix2
+}
+```
+
+Not only is the matrix multiplication a trivially parallelizable operation that likely will be parallelized for any large-ish matrix, but reading the files from disk and then parsing their contents may be done in parallel before the matrix multiplication because the performance benefit is there to do so.
+
+To accomplish this, we need something different from a classic Turing tape to compile to. We call this the Alan Graphcode: a bytecode format with graph annotations that is guaranteed to have no cycles. A dependency graph of operations to perform, with some of these also having a subgraph of operations to perform, provides us with the information we need to know which operations are safe to execute in parallel, coupled with (implied) operation execution time estimates to determine if it makes sense to do so.
```
Comment (depombo) [inline src/the-turing-completeness-problem.md:51]: nit - automatic parallelism within an event handler via AGC and via an array/matrix based on the "size and shape of the data" work differently and I feel like maybe we should tease them apart further?

**Review by depombo — APPROVED** [review]

**Comment by depombo** [comment]
Awesome, exited to start posting the final version of this essay :)

File: src/the-turing-completeness-problem.md (hunk 17 above)
Comment (dfellis) [reply src/the-turing-completeness-problem.md:51]: So I'm also talking about intra-graph parallelization as well, which we aren't planning to tackle for a while. :)

I don't know if teasing them out in this article is worth it, though? In some ways this is similar to the point you made about Duff's Device. Too many topics and concepts all at the same time can detract from the point we're trying to make.

I think explaining our execution model and the advantages it can bring are more important to this article than the gritty details of each particular class of parallelization possible.

**Review by depombo — APPROVED** [review]
